bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringConfiguration.java

```diff
@@ -0,0 +1,157 @@
+      "google.bigtable.mirroring.secondary-client.connection.impl";
+
+  /**
+   * By default all parameters from the Configuration object passed to
+   * ConnectionFactory#createConnection are passed to Connection instances. If this key is set, then
+   * only parameters that start with given prefix are passed to primary Connection. Use it if
+   * primary and secondary connections' configurations share a key that should have different value
+   * passed to each of connections, e.g. zookeeper url.
+   *
+   * <p>Prefixes should not contain dot at the end.
+   */
+  public static final String MIRRORING_PRIMARY_CONFIG_PREFIX_KEY =
+      "google.bigtable.mirroring.primary-client.prefix";
+
+  /**
+   * If this key is set, then only parameters that start with given prefix are passed to secondary
+   * Connection.
+   */
+  public static final String MIRRORING_SECONDARY_CONFIG_PREFIX_KEY =
+      "google.bigtable.mirroring.secondary-client.prefix";
+
+  public MirroringConfiguration(
+      Configuration primaryConfiguration,
+      Configuration secondaryConfiguration,
+      Configuration mirroringConfiguration) {
+    super.set("hbase.client.connection.impl", MirroringConnection.class.getCanonicalName());
+    this.primaryConfiguration = primaryConfiguration;
+    this.secondaryConfiguration = secondaryConfiguration;
+    this.mirroringOptions = new MirroringOptions(mirroringConfiguration);
+  }
+
+  public MirroringConfiguration(Configuration conf) {
+    super(conf); // Copy-constructor
+    // In case the user constructed MirroringConfiguration by hand.
+    if (conf instanceof MirroringConfiguration) {
+      MirroringConfiguration mirroringConfiguration = (MirroringConfiguration) conf;
+      this.primaryConfiguration = new Configuration(mirroringConfiguration.primaryConfiguration);
+      this.secondaryConfiguration =
+          new Configuration(mirroringConfiguration.secondaryConfiguration);
+      this.mirroringOptions = mirroringConfiguration.mirroringOptions;
+    } else {
+      checkParameters(conf);
+      this.primaryConfiguration = constructPrimaryConfiguration(conf);
+      this.secondaryConfiguration = constructSecondaryConfiguration(conf);
+      this.mirroringOptions = new MirroringOptions(conf);
+    }
+  }
+
+  private Configuration constructPrimaryConfiguration(Configuration conf) {
+    return constructConnectionConfiguration(
+        conf, MIRRORING_PRIMARY_CONNECTION_CLASS_KEY, MIRRORING_PRIMARY_CONFIG_PREFIX_KEY);
+  }
+
+  private Configuration constructSecondaryConfiguration(Configuration conf) {
+    return constructConnectionConfiguration(
+        conf, MIRRORING_SECONDARY_CONNECTION_CLASS_KEY, MIRRORING_SECONDARY_CONFIG_PREFIX_KEY);
+  }
+
+  private Configuration constructConnectionConfiguration(
+      Configuration conf, String connectionClassKey, String prefixKey) {
+    String connectionClassName = conf.get(connectionClassKey);
+    String prefix = conf.get(prefixKey, "");
```

Comment:
Why do this instead of hardcoding primary/secondary as the prefix? 

Are you hoping this will help with swapping primary and secondary? That this will remain hbase/bigtable and you will only change the prefix keys to swap the primary database?

Reply:
This is exactly the reason why we have done it. We could have used `bigtable.` and `hbase.` prefixes if we hardcoded them, but we wanted to remain agnostic to underlying implementations and not to make any assumptions that only those two client will be used.